File: java/src/main/java/cn/leetcode/stack/monotone/$456_132Pattern.java
package cn.leetcode.stack.monotone;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * 反序遍历的单减栈，出栈时保留最大值
 */
public class $456_132Pattern {
    public boolean find132pattern(int[] nums) {
        int n = nums.length;
        Deque<Integer> d = new ArrayDeque<>();
        int k = Integer.MIN_VALUE;
        for (int i = n - 1; i >= 0; i--) {
            if (nums[i] < k) return true;
            while (!d.isEmpty() && d.peekLast() < nums[i]) {  // 递减栈
                // 事实上，k 的变化也具有单调性，直接使用 k = pollLast() 也是可以的
                k = Math.max(k, d.pollLast());
            }
            d.addLast(nums[i]);
        }
        return false;
    }

    public static void main(String[] args) {
        boolean ans = new $456_132Pattern().find132pattern(new int[]{3,5,0,3,4});
        System.out.println(ans);
    }
}

File: python/monotone_stack/$456_132Pattern.py
from typing import List


class Solution:
    def find132pattern(self, nums: List[int]) -> bool:
        stack = []
        max_val = -(10 ** 9 + 7)
        for i in range(len(nums) - 1, -1, -1):
            if nums[i] < max_val:  # 最小值哨兵，只要当前元素小于右边出栈元素的最大值，就判定为存在
                return True
            while stack and stack[-1] < nums[i]:  # 大于栈顶元素，维护一个右边比当前值小的最大值的单增栈
                max_val = max(max_val, stack.pop())  # 每次出栈记录最大值
            stack.append(nums[i])
        return False
